crates/goose/src/config/signup_openrouter/server.rs

```diff
@@ -0,0 +1,215 @@
+use anyhow::Result;
+use axum::{
+    extract::Query,
+    http::StatusCode,
+    response::{Html, IntoResponse},
+    routing::get,
+    Router,
+};
+use serde::Deserialize;
+use std::net::SocketAddr;
+use tokio::sync::oneshot;
+
+#[derive(Debug, Deserialize)]
+struct CallbackQuery {
+    code: Option<String>,
+    error: Option<String>,
+}
+
+/// Run the callback server on localhost:3000
+pub async fn run_callback_server(
+    code_tx: oneshot::Sender<String>,
+    shutdown_rx: oneshot::Receiver<()>,
+) -> Result<()> {
+    let app = Router::new().route("/", get(handle_callback));
+
+    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
+
+    let listener = tokio::net::TcpListener::bind(addr).await?;
+
+    // Wrap the code_tx in an Arc<Mutex> so we can use it in the handler
+    let state = std::sync::Arc::new(tokio::sync::Mutex::new(Some(code_tx)));
+
+    axum::serve(listener, app.with_state(state.clone()).into_make_service())
+        .with_graceful_shutdown(async move {
+            let _ = shutdown_rx.await;
+        })
+        .await?;
+
+    Ok(())
+}
+
+async fn handle_callback(
+    Query(params): Query<CallbackQuery>,
+    state: axum::extract::State<
+        std::sync::Arc<tokio::sync::Mutex<Option<oneshot::Sender<String>>>>,
+    >,
+) -> impl IntoResponse {
+    // Check for error first
+    if let Some(error) = params.error {
+        return (
+            StatusCode::BAD_REQUEST,
+            Html(format!(
```

Comment:
so for these html templates, consider using jinja2 for rendering instead of adding a dependency; we could make them external too